lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/NodeObserverIdentifier.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using Org.Apache.REEF.Network.Group.Driver.Impl;
+
+namespace Org.Apache.REEF.Network.Group.Task.Impl
+{
+    /// <summary>
+    /// An identifier for a given node in the group communication graph.
+    /// A node is uniquely identifiable by a combination of its Task ID, 
+    /// <see cref="Type"/>, <see cref="GroupName"/>, and <see cref="OperatorName"/>.
+    /// </summary>
+    internal sealed class NodeObserverIdentifier
+    {
+        private readonly Type _type;
```

Comment:
That will not likely be a problem, since type `T` in `GroupCommunicationMessage` is neither covariant nor contravariant, we cannot downcast/upcast `T` in generic classes. Composition instead of inheritance is the solution to go should the need arise.